algebra/src/curves/mod.rs

```diff
@@ -56,274 +23,106 @@ pub mod tests;
+pub trait Curve:
+    Group
+    + Copy
+    + Serialize
+    + for<'a> Deserialize<'a>
+    + UniformRand
+    + From<<Self as Curve>::AffineRep>
+    + TryInto<<Self as Curve>::AffineRep, Error = Error>
+{
+    type BaseField: Field + SquareRootField;
+    type AffineRep: Sized
         + Sync
+        + Copy
+        + PartialEq
         + Debug
-        + From<Self::G1Affine>;
-
-    /// The projective representation of an element in G2.
-    type G2Projective: ProjectiveCurve<BaseField = Self::Fqe, ScalarField = Self::Fr, Affine = Self::G2Affine>
-        + From<Self::G2Affine>
-        + Into<Self::G2Affine>;
-
-    /// The affine representation of an element in G2.
-    type G2Affine: AffineCurve<BaseField = Self::Fqe, ScalarField = Self::Fr, Projective = Self::G2Projective>
-        + From<Self::G2Projective>
-        + Into<Self::G2Projective>
-        + Into<Self::G2Prepared>;
-
-    /// A G2 element that has been preprocessed for use in a pairing.
-    type G2Prepared: ToBytes
-        + FromBytes
+        + Hash
         + Serialize
         + for<'a> Deserialize<'a>
-        + Default
-        + Eq
-        + PartialEq
-        + Clone
-        + Send
-        + Sync
-        + Debug
-        + From<Self::G2Affine>;
-
-    /// The base field that hosts G1.
-    type Fq: PrimeField + SquareRootField;
-
-    /// The extension field that hosts G2.
-    type Fqe: SquareRootField;
-
-    /// The extension field that hosts the target group of the pairing.
-    type Fqk: Field;
-
-    /// Perform a miller loop with some number of (G1, G2) pairs.
-    #[must_use]
-    fn miller_loop<'a, I>(i: I) -> Result<Self::Fqk, Error>
-    where
-        I: IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>;
-
-    /// Perform final exponentiation of the result of a miller loop.
-    #[must_use]
-    fn final_exponentiation(_: &Self::Fqk) -> Result<Self::Fqk, Error>;
+        + CanonicalSerialize
+        + CanonicalDeserialize
+        + ToBytes
+        + FromBytes
+        + TryFrom<Self, Error = Error>;
 
-    /// Computes a product of pairings.
-    #[must_use]
-    fn product_of_pairings<'a, I>(i: I) -> Result<Self::Fqk, Error>
-    where
-        I: IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>,
-    {
-        Self::final_exponentiation(&Self::miller_loop(i)?)
+    #[inline]
+    fn into_affine(&self) -> Result<Self::AffineRep, Error> {
+        TryInto::<Self::AffineRep>::try_into(*self)
     }
 
-    /// Performs multiple pairing operations
-    #[must_use]
-    fn pairing<G1, G2>(p: G1, q: G2) -> Result<Self::Fqk, Error>
-    where
-        G1: Into<Self::G1Affine>,
-        G2: Into<Self::G2Affine>,
-    {
-        let g1_prep = Self::G1Prepared::from(p.into());
-        let g2_prep = Self::G2Prepared::from(q.into());
-        Self::product_of_pairings(std::iter::once(&(g1_prep, g2_prep)))
+    #[inline]
+    fn from_affine<'a>(other: &'a Self::AffineRep) -> Self {
+        Into::<Self>::into(*other)
     }
```

Comment:
Are these actually needed ? We already have the TryFrom traits for both Curve and AffineRep

Reply:
This is just for more laconic calls. Rust compiler often cannot infer type when we use something like

a_affine.into()

Specifying needed types make the code construction too complex and unreadable. In general - yes, we can avoid to use this "from_affine" function